adding video compression (h264)

bsobhani opened the pull request:
Starting pull request for video compression (h264) support. Not close to being perfect, but getting the process started so that it does not linger.

Basic functionality works and seems stable on the machines I tested it on. Achieves lower compression sizes with higher quality than existing codecs for some types of image sequences. For example, if the differences between frames is consistently small, like at SNS beamlines where neutron events accumulate on an image.

To compile, set:

WITH_VC=YES
VC_EXTERNAL=NO

To use, select "H264" from $(P)$(R)Compressor.

Tuning PVs:
* $(P)$(R)QMinMax - 1 for least lossless compression, higher values for lossier compression
* $(P)$(R)GOPSize - Group of picture size. 1 for frame-by-frame compression. Set to higher values for true video compression.

bsobhani committed:
adding video compression (h264)

ericonr commented on supportApp/videoCompressionSrc/README:
```diff
@@ -0,0 +1,7 @@
+Prerequisites:
+* ffmpeg needs to be installed. "yum install ffmpeg-devel"
+
+PVs:
+* $(P)$(R)QMinMax - 1 for least lossless compression, higher values for lossier compression
+* $(P)$(R)GOPSize - Group of picture size. 1 for frame-by-frame compression. Set to higher values for true video compression.
+
```
Doesn't this PV documentation belong in NDPluginCodec?

And "1 for least *lossy* compression"

ericonr commented on supportApp/videoCompressionSrc/codec_tools.c:
```diff
@@ -0,0 +1,222 @@
+//#include "pgm.h"
+#include "codec_tools.h"
+
+AVCodecContext *c_c = 0; 
+int first = 1;
+AVCodecContext *c_d = 0; 
+int count_d = 0;
+
+
+const AVCodec *codec;
```
Wouldn't it be best to avoid global variables? Doesn't this mean a single IOC can only have a single compressed stream?

ericonr commented on supportApp/videoCompressionSrc/codec_tools.c:
```diff
@@ -0,0 +1,222 @@
+	avcodec_register_all();
+	codec = avcodec_find_encoder(AV_CODEC_ID_H264);
+	c = avcodec_alloc_context3(codec);
+	c->bit_rate = 400000;
+	c->width = width;
+	c->height = height;
+	//c->qmin = 1;
+	//c->qmax = 20;
+	c->time_base = (AVRational){1, 25};
+	c->framerate = (AVRational){25, 1};
+	//av_opt_set(c->priv_data, "crf", "1", AV_OPT_SEARCH_CHILDREN);
+	//av_opt_set(c->priv_data, "crf", "0", 0);
+	//av_opt_set(c->priv_data, "crf", "20", 0);
+	c->gop_size = 100;
```
Some of these should be configurable by the users as well, no?

ericonr replied:
I feel like `bit_rate`, `time_base` and `framerate` at least should be :thinking:

ericonr commented on supportApp/videoCompressionSrc/codec_tools.c:
```diff
@@ -0,0 +1,222 @@
+
+	int ret = avcodec_open2(c, codec, NULL);
+
+	if (ret < 0) {
+		fprintf(stderr, "Could not open codec: %s\n", av_err2str(ret));
+		exit(1);
+	}
```
Failures in library code shouldn't cause the IOC to exit, right?

ericonr commented on supportApp/videoCompressionSrc/video_compression.c:
```diff
@@ -0,0 +1,28 @@
+	re_init_encoder_context();
+}
+
+int H264_decompress(const char* source, char* dest, int originalSize)
+{
+
+        int width, height;
+        decompress_buffer(source, originalSize, dest, &width, &height);
+	return width*height;
+
+}
```
This function isn't declared in the header.

ericonr replied:
I see now that it's used in ADViewer. IMO it still should be in the header, as an exported interface.

ericonr commented on supportApp/videoCompressionSrc/video_compression.h:
```diff
@@ -0,0 +1,20 @@
+extern "C" {
+#endif
+
+int H264_compress(const char* source, char* dest, int x_size, int y_size);
```
Shouldn't the size for `dest` be passed to this function?

I'd also feel more comfortable with the type in use being `unsigned char` rather than `char`. Pixels won't have negative values after all.

bsobhani replied:
dest is the buffer where the compressed image is stored, but the buffer itself is given the same size as the uncompressed image. H264_compress returns the number of bytes written to dest. If the compressed image is for some reason larger than the uncompressed image, the current behavior is to consider it a failure and reject the compressed image.

bsobhani replied:
The only reason H264_compress would need the size of dest would be if it did the checking to see if the dest buffer is large enough. The H264_compress function is currently implemented under the assumption that the dest buffer is large enough to store the compressed image. Whether it is large enough depends on the upper bound function being accurate enough (which I admit this is unknown to me), since this is what ensures that dest is large enough. So far I have not noticed issues here, but I will do testing to further investigate this.

ericonr reviewed (commented):
The code and Makefiles still have a lot of comments and debugging printfs...

bsobhani commented:
Erico, thank you for the feedback. I have responded to some of your comments. When I have time I will take a more thorough look at the pull request and probably make some additional commits.

bsobhani committed:
add lock/unlock for parameter writing, so that H264_compress can be called in ADCore without lock

bsobhani committed:
create CodecContext struct that extends AVCodec context, and modify some functions to take CodecContext object as a parameter

bsobhani committed:
make it so that compression function and param setter functions take context object as parameter instead of global context object. Also, make interfaces to these functions that take context as void pointer so that files that call these functions do not need to include libav headers

bsobhani committed:
make depreciated av_register function conditional upon libavcodec version, for compatibility.
yuv422 seems to have issues for certain setups (openh264). change to yuv420 until further investigation
made set_param_int function

bsobhani committed:
comment out global mutex and declare context initialization function

bsobhani committed:
use correct context initialization function

MarkRivers commented:
What is the status of this PR?  Have all of @ericonr comments been addressed?

bsobhani commented:
Hi Mark, sorry the delay, this has actually been on my mind lately. I have a couple of small commits from the summer which I did not push for the sake of not wanting to distract anyone. I will aim to address the remainder of erico's comments this weekend. I believe they would be small changes.

ericonr commented:
@MarkRivers this PR was discussed a few times during the AreaDetector collab meetings.

We noticed a few conceptual issues:

- if a PVA client connects in the middle of the h264 stream, it doesn't have a way to signal for a key frame, so it will have to wait for one in order to display the image
- codecs might behave weirdly when dealing with variable frame-rate
- does this provide a considerably better experience than using ffmepgServer, for example?

bsobhani commented:
Erico, I do not see why (1) would matter except in strange circumstances that one may be able to construct. Can you explain what you mean by (2)? For (3) the advantage is it can be displayed in a PVA client.

ericonr commented:
About (1), how not? If an IOC is up, in the middle of an acquisition, and someone opens up an interface to monitor its progress, the data stream they'll be receiving won't have the necessary information to display frames until there is a key frame. Given how PVA works, it can also happen that some intermediate update is omitted from the data stream; how would a client deal with such a corrupted stream?

About (2), it has to do with passing the h264 stream to other viewers, which might have specific requirements for video streams (display at x Hz), instead of supporting playback of "each frame as it arrives".

About (3), I can see how that can be an advantage, but I also feel like the array normative type wasn't designed for interframe compression, and fitting support into it seems like it would require more than just this PR (designers of the type feel free to correct me); if one simply uses ffmpegServer, or intraframe compression (e.g. JPEG), similar results can be obtained.

bsobhani commented:
I do not see (1) as important because the amount of time is negligible and I do not understand (2) or (3). If you would rather not have this merged please feel free to close this pull request.

ericonr commented:
@bsobhani I am not Mark, so I don't have the power to decide on such things. I merely wanted to point out what had been discussed.

The amount of time for key frames being negligible seems pretty relevant and like a positive point for your proposal. It might also make (3) less relevant. I don't have any experience with using this PR, so I wouldn't know.

Therefore, I don't think preemptively closing the PR is necessary.
